docs/fileformat-prose.md

```diff
@@ -0,0 +1,318 @@
+but are [easy to find].
+```
+
+We represent a sparse array as a `Group` on-disk,
+where the kind and shape of the sparse array is defined in the `Group`'s attributes:
+
+```python
+>>> dict(f["X"].attrs)
+{'encoding-type': 'csr_matrix',
+ 'encoding-version': '0.1.0',
+ 'shape': array([38410, 27899])}
+```
+
+The group contains three arrays:
+
+```python
+>>> f["X"].visititems(print)
+data <HDF5 dataset "data": shape (41459314,), type "<f4">
+indices <HDF5 dataset "indices": shape (41459314,), type "<i4">
+indptr <HDF5 dataset "indptr": shape (38411,), type "<i4">
+```
+
+<!-- ```bash
+$ h5ls 02_processed.h5ad/X
+data                     Dataset {41459314/Inf}
+indices                  Dataset {41459314/Inf}
+indptr                   Dataset {38411/Inf}
+``` -->
+
+### Sparse array specification (v0.1.0)
+
+* Each sparse array MUST be its own group
+* The group MUST contain arrays `indices`, `indptr`, and `data`
+* The group's metadata MUST contain:
+    * `"encoding-type"`, which is set to `"csr_matrix"` or `"csc_matrix"` for compressed sparse row and compressed sparse column, respectively.
+    * `"encoding-version"`, which is set to `"0.1.0"`
+    * `"shape"` which is an integer array of length 2 whose values are the sizes of the array's dimensions
+
+## DataFrames
+
+DataFrames are saved as a columnar format in a group, so each column of a DataFrame is saved as a separate array.
+We save a little more information in the attributes here.
+
+```python
+>>> dict(f["obs"].attrs)
+{'_index': 'Cell',
+ 'column-order': array(['sample', 'cell_type', 'n_genes_by_counts',
+        'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts',
+        'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes',
+        'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes',
+        'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito',
+        'label_by_score'], dtype=object),
+ 'encoding-type': 'dataframe',
+ 'encoding-version': '0.2.0'}
+```
+
+These attributes identify the index of the dataframe, as well as the original order of the columns.
+Each column in this dataframe is encoded as its own array.
+
+```python
+>>> dict(f["obs"]["total_counts"].attrs)
+{'encoding-type': 'array', 'encoding-version': '0.2.0'}
+
+>>> dict(f["obs"]["cell_type"].attrs)
+{'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
+```
+
+### Dataframe Specification (v0.2.0)
+
+* A dataframe MUST be stored as a group
+* The group's metadata:
+    * MUST contain the field `"_index"`, which is the name of the column to be used as an index
+    * MUST contain encoding metadata `"encoding-type: "dataframe"`, `"encoding-version": "0.2.0"`
+    * MUST contain `"column-order"` an array of strings denoting the order of column entries
+* The group MUST contain a column for the index
+* Each entry in the group MUST correspond to an array with equivalent first dimensions
```

Comment:
```suggestion
### Dataframe Specification (v0.2.0)

* A dataframe MUST be stored as a group
* The group's metadata:
    * MUST contain the field `"_index"`, whose value is the key of the array to be used as an index
    * MUST contain encoding metadata `"encoding-type": "dataframe"`, `"encoding-version": "0.2.0"`
    * MUST contain `"column-order"` an array of strings denoting the order of column entries
* The group MUST contain an array for the index
* The index array MUST NOT appear in 
* Each entry in the group MUST correspond to an array with equivalent first dimensions
```

Reply:
I think that's good apart from the missing end of the sentence.

Just for my understanding where is the index stored if it's not a column? Is it still in the array but named something else? Or should the sentence say something like "MUST contain the field `"_index"`, whose value is the key of the _group containing an array_ to be used as an index? It's probably obvious, just my lack of familiarly with HDF5.

Reply:
> I think that's good apart from the missing end of the sentence.

Oh, I think I meant to actually remove that bullet, since that is allowed (maybe it shouldn't be)

> Just for my understanding where is the index stored if it's not a column?

It's still stored in the group. For example:

```python
>>> for k, v in f["obs"].items(): print(k, ":", dict(v.attrs))
Cell : {'encoding-type': 'string-array', 'encoding-version': '0.2.0'}
cell_type : {'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
label_by_score : {'encoding-type': 'categorical', 'encoding-version': '0.2.0', 'ordered': False}
log1p_n_genes_by_counts : {'encoding-type': 'array', 'encoding-version': '0.2.0'}
...
```

It's just not included as a column of the dataframe since it's not included in `f["obs"].attrs["column-order"]`.

One could have an array be both an index and a value inside the dataframe. This is actually probably a nice organizational practice. However, this could be a problem if you instead have an index and a column with the same names, but different values.